README.md
# eror

[![NPM version][meta-img-npm]][meta-url-npm]
[![Build status][meta-img-travis]][meta-url-travis]
[![Coveralls status][meta-img-coveralls]][meta-url-coveralls]
[![Support us][meta-img-gratipay]][meta-url-gratipay]

Error utilities which provide error classes for HTTP response (4xx, 5xx) and more.

## Installation

```sh
$ npm install eror
```

## Example

```js
// Formatted message
new eror.InvalidTypeError('"%s" must be a string!', 'name');

// Custom status code
new eror.MissingValueError(400, '"%s" is required!', 'age');

// Pre-defined HTTP errors
new eror.NotFoundError('User "%s" not found!', 'paul');

// Custom errors
eror(503, 'The Earth is %s!', 'exploding');
```

## API reference

### eror([statusCode], [message || format], [...values])

Create a custom error.

```js
eror(401, 'You are not logged in!');
eror('Welcome, %s %s.', 'John', 'Lennon');
```

* `statusCode`: Status code as a number, defaults to 500.
* `message`: Error message.
* `format`: Placeholder string, please refer to
[util.format][url-nodejs-util-format] for more details.
* `...values`: Values to be placed in placeholder string.

### new eror\[name\]([statusCode], [message || format], [...values])

Instantiate a new common error.

```js
new eror.MissingValueError(400, '"%s" field is required!', 'name');
```

* `name`: Class name of the error, as specified in [common error list](#common-errors).
* `statusCode`: Status code, if is not specified, default status code, 500, will be used.
* `message`, `format` and `...values`: as in [eror()](#erorstatuscode-message--format-values).

### new eror\[name || statusCode\]([message || format], [...values])

Instantiate a new HTTP error.

```js
new eror.NotFoundError('Requested resource not found!');
new eror[500]('Server exploded at %s!', new Date());
```

* `name`: Class name of the error, as specified in [HTTP error list](#http-errors).
* `statusCode`: Status code of HTTP error.
* `message`, `format` and `...values`: as in [eror()](#erorstatuscode-message--format-values).

HTTP error's status code cannot be overridden.

### Error properties

```js
var err = eror(404, 'Nothing!');

err.message; // Nothing!
err.statusCode; // 404
err.status; // also 404
```

* `message`
* `statusCode` and `status`: Status code of the error.

### Error class properties

```js
eror.InternalServerError.MESSAGE; // Internal Server Error
eror.InternalServerError.STATUS_CODE; // 500
eror.InternalServerError.STATUS; // also 500
```

* `MESSAGE`: Default error message of errors which are instantiate from this class.
* `STATUS_CODE` and `STATUS`: Default status code.

### Error list

#### Common errors

| Class name                | Status code |
| ------------------------- |:-----------:|
| UnsupportedOperationError |     500     |
| ReversedNameError         |     500     |
| NameAlreadyDefinedError   |     500     |
| InvalidOperationError     |     500     |
| InvalidTypeError          |     500     |
| MissingValueError         |     500     |
| InvalidValueError         |     500     |

#### HTTP errors

| Class name                         | Status code |
| ---------------------------------- |:-----------:|
| BadRequestError                    |     400     |
| UnauthorizedError                  |     401     |
| PaymentRequiredError               |     402     |
| ForbiddenError                     |     403     |
| NotFoundError                      |     404     |
| MethodNotAllowedError              |     405     |
| NotAcceptableError                 |     406     |
| ProxyAuthenticationRequiredError   |     407     |
| RequestTimeoutError                |     408     |
| ConflictError                      |     409     |
| GoneError                          |     410     |
| LengthRequiredError                |     411     |
| PreconditionFailedError            |     412     |
| PayloadTooLargeError               |     413     |
| UriTooLongError                    |     414     |
| UnsupportedMediaTypeError          |     415     |
| RangeNotSatisfiableError           |     416     |
| ExpectationFailedError             |     417     |
| ImATeapotError                     |     418     |
| UnprocessableEntityError           |     422     |
| LockedError                        |     423     |
| FailedDependencyError              |     424     |
| UnorderedCollectionError           |     425     |
| UpgradeRequiredError               |     426     |
| PreconditionRequiredError          |     428     |
| TooManyRequestsError               |     429     |
| RequestHeaderFieldsTooLargeError   |     431     |
| UnableForLegalReasonsError         |     451     |
| InternalServerError                |     500     |
| NotImplementedError                |     501     |
| BadGatewayError                    |     502     |
| ServiceUnavailableError            |     503     |
| GatewayTimeoutError                |     504     |
| HttpVersionNotSupportedError       |     505     |
| VariantAlsoNegotiatesError         |     506     |
| InsufficientStorageError           |     507     |
| LoopDetectedError                  |     508     |
| BandwidthLimitExceededError        |     509     |
| NotExtendedError                   |     510     |
| NetworkAuthenticationRequiredError |     511     |

## Contributing

* [Submit new issue][repo-url-new-issue] if there is any common error, which is
useful for you or others, unavailable.

* In lieu of a formal styleguide, take care to maintain the existing coding style.
  [.editorconfig][repo-editorconfig] and [.jshintrc][repo-jshintrc] files can give you
  more details.

* Add tests for any new or changed functionality/element.

* Lint and test your code using `npm test`.

### Author

* Website: [http://meo.guru][url-meoguru-website]
* Twitter: [@meoguru][url-meoguru-twitter]

## License

Released under [MIT license][repo-license].

[//]: # (Site URLs)
[url-editorconfig]: http://editorconfig.org
[url-jshint-docs]: http://www.jshint.com/docs
[url-nodejs-util-format]: http://nodejs.org/api/util.html#util_util_format_format

[//]: # (Repository URLs and resources)
[repo-url-new-issue]: https://github.com/meoguru/eror/issues/new
[repo-url-pull-request]: https://github.com/meoguru/eror/pulls
[repo-license]: https://github.com/meoguru/eror/blob/master/LICENSE
[repo-editorconfig]: https://github.com/meoguru/eror/blob/master/.editorconfig
[repo-jshintrc]: https://github.com/meoguru/eror/blob/master/.jshintrc
[repo-examples]: https://github.com/meoguru/eror/tree/master/examples

[//]: # (Repository meta information)
[meta-url-npm]: https://npmjs.org/package/eror
[meta-img-npm]: https://img.shields.io/npm/v/eror.svg?style=flat
[meta-url-travis]: https://travis-ci.org/meoguru/eror
[meta-img-travis]: https://img.shields.io/travis/meoguru/eror.svg?style=flat
[meta-url-coveralls]: https://coveralls.io/r/meoguru/eror
[meta-img-coveralls]: https://img.shields.io/coveralls/meoguru/eror/master.svg?style=flat
[meta-url-gratipay]: https://gratipay.com/meoguru
[meta-img-gratipay]: https://img.shields.io/gratipay/meoguru.svg?style=flat

[//]: # (Author URLs)
[url-meoguru-website]: http://meo.guru
[url-meoguru-twitter]: http://twitter.com/meoguru


CHANGELOG.md
# 1.0.0 (2014-11-26)

Features:

* Allow error message formatting.
* Suport 7 common errors and 40 HTTP errors.
* Suport custom errors.


test/lib/util.js
'use strict';

exports.assign = function (dest, src) {
  for (var prop in src) {
    if (src.hasOwnProperty(prop)) {
      dest[prop] = src[prop];
    }
  }
};


test/lib/setup.js
'use strict';

var chai = require('chai');
var sinon = require('sinon'); // jshint ignore: line
var sinonChai = require('sinon-chai');

chai.use(sinonChai);

global.expect = chai.expect;
global.sinon = sinon;


test/unit/eror.js
'use strict';

var eror = require('../..');

var commonErrors = require('./error/common');
var httpErrors = require('./error/http');

var util = require('../lib/util');

var errors = {};

util.assign(errors, commonErrors);
util.assign(errors, httpErrors);

describe('eror', function () {
  it('should create custom error', function () {
    var err = eror();
    expect(err).to.be.instanceOf(Error);
  });

  it('should use default status code of 500', function () {
    var err = eror('Boom!');
    expect(err).to.have.property('statusCode', 500);
    expect(err).to.have.property('status', 500);
  });

  it('should allow overriding status code', function () {
    var err = eror(499);
    expect(err).to.have.property('statusCode', 499);
  });

  it('should not allow out-of-range status code', function () {
    expect(function () {
      eror(399);
    }).to.throw(Error);
  });

  it('should format error message', function () {
    var err = eror('Hello %s %d %j!', 'world', 3.14, { foo: 'bar' });
    expect(err).to.have.property('message', 'Hello world 3.14 {"foo":"bar"}!');
  });

  for (var name in errors) {
    if (!errors.hasOwnProperty(name)) {
      continue;
    }

    (function (name) {
      describe('.' + name, function () {
        it('should be a valid error class', function () {
          var err = new eror[name]();
          expect(err).to.be.instanceOf(eror[name]);
          expect(err).to.be.instanceOf(Error);
        });

        it('should have default status code', function () {
          expect(eror[name]).to.have.property('STATUS_CODE', errors[name].statusCode);
          expect(eror[name]).to.have.property('STATUS', errors[name].statusCode);
        });

        it('should have default error message', function () {
          expect(eror[name]).to.have.property('MESSAGE', errors[name].message);
        });

        it('should use default status code if not set', function () {
          var err = new eror[name]();
          expect(err).to.have.property('statusCode', eror[name].STATUS_CODE);
          expect(err).to.have.property('status', eror[name].STATUS_CODE);
        });

        it('should use default error message if not set', function () {
          var err = new eror[name]();
          expect(err).to.have.property('message', eror[name].MESSAGE);
        });

        if (commonErrors[name]) {
          it('should allow overriding status code', function () {
            var err = new eror[name](499);
            expect(err).to.have.property('statusCode', 499);
          });

          it('should not allow out-of-range status code', function () {
            expect(function () {
              new eror[name](399); // jshint ignore: line
            }).to.throw(Error);
          });

        } else {
          it('should not allow overriding status code', function () {
            expect(function () {
              new eror[name](499); // jshint ignore: line
            }).to.throw(Error);
          });
        }

        it('should format error message', function () {
          var err = new eror[name]('Hello %s %d %j!', 'world', 3.14, { foo: 'bar' });
          expect(err).to.have.property('message', 'Hello world 3.14 {"foo":"bar"}!');
        });
      });
    })(name); // jshint ignore: line
  }
});


gulpfile.js
'use strict';

var gulp = require('gulp');
var $ = require('gulp-load-plugins')();

var expand = require('expands');

/* Constants */

var constant = expand({
  dir: {
    lib: 'lib',
    test: 'test',
    testLib: '{dir.test}/lib',
    unitTest: '{dir.test}/unit',
    testResources: '{dir.test}/resources',
  },
  file: {
    testSetup: '{dir.testLib}/setup.js'
  },
  set: {
    all: [
      '{dir.lib}/**/*.js',
      '{dir.test}/**/*.js',
      'gulpfile.js'
    ],
    code: [ '{dir.lib}/**/*.js' ],
    test: [ '{dir.test}/**/*.js' ],
    testSets: [
      '{dir.test}/**/*.js',
      '!{dir.testLib}/**/*.js',
      '!{dir.testResources}/**/*.js'
    ],
    unitTests: [ '{dir.unitTest}/**/*.js' ],
    testResources: [ '{dir.testResources}/**/*.js' ],
    testTriggers: [
      '{dir.lib}/**/*.js',
      '{dir.testLib}/**/*.js',
      '{dir.testResources}/**/*.js'
    ]
  }
});

var f = constant.file;
var s = constant.set;

/* Gulp helpers */

var watching = false;

gulp.doneCallback = function (err) {
  if (!watching) {
    process.exit(err ? 1 : 0);
  }
};

var onError = function (err) {
  $.util.log(
    $.util.colors.red(
      err && err.stack ? err.stack : err
    )
  );

  if (watching) {
    this.emit('end');
  }
  else {
    process.exit(1);
  }
};

/* Gulp tasks */

gulp.task('lint', function () {
  return lint(
    gulp.src(s.all)
  );
});

gulp.task('unit-test', function () {
  return test(
    gulp.src(s.unitTests)
  );
});

gulp.task('test', function (cb) {
  gulp.src(s.code)
    .pipe($.istanbul({
      includeUntested: true
    }))
    .on('error', $.notify.onError({
      title: 'Test',
      message: '<%= error.message %>'
    }))
    .on('error', onError)
    .on('finish', function () {
      test(gulp.src(s.testSets))
        .pipe($.istanbul.writeReports())
        .on('end', cb);
    });
});

gulp.task('watch', function () {
  watching = true;

  // Lints changed files
  $.watch(s.all, lint);

  // Tests using only changed test files
  $.watch(s.testSets, test);

  // Executes unit tests on code and test resource changes
  gulp.watch(s.testTriggers, [ 'unit-test' ]);
});

gulp.task('coveralls', [ 'lint', 'test' ], function () {
  return gulp.src('coverage/lcov.info')
    .pipe($.coveralls());
});

gulp.task('default', [ 'watch' ]);

/* Gulp task logic */

function lint(src) {
  return src
    .pipe($.jshint())
    .pipe($.jshint.reporter('jshint-stylish'))
    .pipe($.notify({
      title: 'Lint',
      message: function (file) {
        if (!file.jshint || file.jshint.success) {
          return false;
        }

        var errors = file.jshint.results.map(function (data) {
          if (data.error) {
            return $.util.template(
              '(<%= line %>:<%= character %>) <%= reason %>',
              {
                file: file,
                line: data.error.line,
                character: data.error.character,
                reason: data.error.reason
              }
            );
          }
        }).join('\n');

        return '<%= file.relative %> ' +
          '(<%= file.jshint.results.length %> ' +
          'error<% if (file.jshint.results.length > 1) { %>s<% } %>).\n' +
          errors;
      }
    }))
    .pipe($.jshint.reporter('fail'))
    .on('error', onError);
}

function test(src) {
  require('./' + f.testSetup);

  return src
    .pipe($.mocha({ reporter: 'dot' }))
    .on('error', $.notify.onError({
      title: 'Test',
      message: '<%= error.message %>'
    }))
    .on('error', onError);
}


lib/eror.js
'use strict';

var inherits = require('util').inherits;
var format = require('util').format;

var commonErrors = require('./error/common');
var httpErrors = require('./error/http');

var util = require('./util');

var slice = [].slice;

var eror = module.exports = function (statusCode) {
  var args = slice.call(arguments);

  if (util.isNumber(statusCode)) {
    statusCode = ~~statusCode;
    args.shift();

  } else {
    statusCode = 500;
  }

  var message = format.apply(null, args);
  var err = create(statusCode, message);

  Error.captureStackTrace(err, Error);

  return err;
};

commonErrors.forEach(function (message) {
  expose(500, message);
});

for (var statusCode in httpErrors) {
  /* istanbul ignore next */
  if (httpErrors.hasOwnProperty(statusCode)) {
    expose(Number(statusCode), httpErrors[statusCode]);
  }
}

function expose(defaultStatusCode, defaultMessage) {
  var name = defaultMessage.replace(/([\w']+)(\s|$)/g, function (all, word) {
    word = word.replace(/[^\w]/g, '');
    return word.charAt(0).toUpperCase() + word.substring(1).toLowerCase();
  });

  if (!util.endsWith(name, 'Error')) {
    name = name + 'Error';
  }

  var allowStatusCodeOverriding = (httpErrors[defaultStatusCode] !== defaultMessage);

  eror[name] = function (statusCode, message) {
    var args = slice.call(arguments);

    if (util.isNumber(statusCode)) {
      statusCode = ~~statusCode;

      if (!allowStatusCodeOverriding && statusCode !== defaultStatusCode) {
        throw eror('Status code of HTTP error cannot be overridden!');
      }

      args.shift();

    } else {
      message = statusCode;
      statusCode = defaultStatusCode;
    }

    if (message) {
      message = format.apply(null, args);
    } else {
      message = defaultMessage;
    }

    var err = create(statusCode, message);

    err.__proto__ = eror[name].prototype;
    Error.captureStackTrace(err, eror[name]);

    return err;
  };

  eror[name].STATUS_CODE = defaultStatusCode;
  eror[name].MESSAGE = defaultMessage;

  util.alias(eror[name], 'STATUS', 'STATUS_CODE');

  inherits(eror[name], Error);
}

function create(statusCode, message) {
  if (statusCode < 400 || statusCode > 599) {
    throw eror('Bad status code, must be in range of 400-599!');
  }

  var err = new Error();

  err.statusCode = statusCode;
  err.message = message;

  util.alias(err, 'status', 'statusCode');

  return err;
}
